[xray] Implements ray.wait

elibol opened the pull request:
Implements `ray.wait` for xray via the local scheduler. This includes both back-end and front-end changes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5712/
Test FAILed.

ericl commented:
Btw does this also fix https://github.com/ray-project/ray/issues/1128 ?

elibol commented:
Chatted with @ericl offline: The existing semantics of `num_returns` is to return **at most** the number of object ids specified by `num_returns`. This implementation changes the semantics to be return **at least** the number of object ids specified by `num_returns`. For instance, if you do `ray.wait` on 100 objects with `num_returns=2`, and 50 objects are done, then `ray.wait` returns 50 objects instead of just 2.

Here are some options:
1. Change the semantics of `num_returns`, which solves #1128. @ericl notes this breaks syntax such as `[x,y] = ray.wait(object_ids, num_returns=2)`.
2. Keep existing semantics, but come up with a way to be "fair" with which objects are returned. Some ideas include random sampling or stratified sampling over nodes. We should include a test that captures these semantics.

@robertnishihara can you provide your thoughts on this?

stephanie-wang commented:
I think we should keep the existing semantics since changing it now would be a surprise to other Ray users.

I'm not sure if "fairness" is necessary for `ray.wait` semantics? I would be okay with just taking the first `num_returns` objects, in whatever order the object manager happens to process them in.

stephanie-wang commented on src/ray/object_manager/object_manager.cc:
```diff
@@ -306,17 +307,119 @@ ray::Status ObjectManager::SendObjectData(const ObjectID &object_id,
+  }
+
+  RAY_CHECK(object_ids.size() != 0);
+  RAY_CHECK(num_required_objects != 0);
```
Same here, we could return `[], object_ids` immediately.

stephanie-wang commented on src/ray/object_manager/object_manager.cc:
```diff
@@ -306,17 +307,119 @@ ray::Status ObjectManager::SendObjectData(const ObjectID &object_id,
+  // will do nothing on non-zero error codes.
+  wait_state.timeout_timer->cancel();
+  // Invoke the wait handler.
+  int64_t time_taken =
```
What is this used for?

stephanie-wang commented on src/ray/object_manager/object_directory.cc:
```diff
@@ -86,25 +98,65 @@ ray::Status ObjectDirectory::GetInformation(const ClientID &client_id,
+    status = gcs_client_->object_table().RequestNotifications(
+        JobID::nil(), object_id, gcs_client_->client_table().GetLocalClientId());
+  }
+  if (listeners_[object_id].callbacks.count(callback_id) > 0) {
```
I think we should probably make this a fatal check, unless there is a valid case where this would happen.

elibol replied:
The only case I can think of is when a `Pull` is invoked multiple times for the same object. I think we should keep this the way it is until we implement `Pull` suppression (I've added a TODO). If we make this fatal now, we may cause unnecessary failures.

stephanie-wang commented on src/ray/object_manager/object_manager.cc:
```diff
@@ -306,17 +307,119 @@ ray::Status ObjectManager::SendObjectData(const ObjectID &object_id,
+                wait_state.found.insert(object_id);
+              }
+              wait_state.requested_objects.erase(object_id);
+              if (wait_state.requested_objects.empty()) {
```
Why not return on the same check of `(wait_state.found.size() >= wait_state.num_required_objects)`? If it's because you're waiting for all of the results to return from the object table, that's probably not necessary. If we can, let's try to keep the logic the same between the `Lookup` and the `Subscribe` codepaths. It'd be best if we could use a common helper function between the two.

Also, if we change that check, we could probably get rid of the `requested_objects` member.

elibol replied:
We ought to keep `requested_objects` as this makes the code easier to read and reason about. Let's talk offline about the other changes.

stephanie-wang replied:
Okay, I'm fine with keeping `requested_objects`, but please add a comment for why you're completing the `Wait` here, since it's not obvious to the reader why the condition is different from the `Subscribe` codepath.

stephanie-wang commented on src/ray/object_manager/object_manager.h:
```diff
@@ -194,6 +195,30 @@ class ObjectManager : public ObjectManagerInterface {
+  UniqueID object_directory_pull_callback_id_ = UniqueID::from_random();
+
+  struct WaitState {
+    WaitState(asio::io_service &service, int64_t wait_ms, const WaitCallback &callback) :
```
Please document the members of this struct.

stephanie-wang commented on src/ray/object_manager/object_manager.h:
```diff
@@ -194,6 +195,30 @@ class ObjectManager : public ObjectManagerInterface {
+        callback(callback)
+    {}
+    int64_t wait_ms;
+    std::shared_ptr<boost::asio::deadline_timer> timeout_timer;
```
Why `shared_ptr` instead of `unique_ptr`?

robertnishihara commented:
Yeah, I think the existing semantics make sense, especially since `ray.wait` is often used to wait for exactly 1 object, and so having to handle more than 1 object would be confusing.

If you want to just get all of the objects that are ready after a certain timeout, then you can specify a timeout along with `num_returns=len(object_ids)`.

robertnishihara commented on test/runtest.py:
```diff
@@ -758,9 +758,6 @@ def testGetMultiple(self):
 
-    @unittest.skipIf(
-        os.environ.get("RAY_USE_XRAY") == "1",
-        "This test does not work with xray yet.")
```
I think there are other tests that can be added back in, e.g., `testMultipleWaitsAndGets` and `testWait` in `stress_tests.py`.

robertnishihara commented on python/ray/worker.py:
```diff
@@ -2574,18 +2571,32 @@ def wait(object_ids, num_returns=1, timeout=None, worker=global_worker):
+            if num_returns > len(object_ids):
+                raise Exception("num_returns cannot be greater than the number "
+                                "of objects provided to ray.wait.")
+            timeout = timeout if timeout is not None else 2**30
```
all the above code should run in both code paths, not just the raylet code path

robertnishihara commented on python/ray/worker.py:
```diff
@@ -2574,18 +2571,32 @@ def wait(object_ids, num_returns=1, timeout=None, worker=global_worker):
+            object_id_strs = [
+                plasma.ObjectID(object_id.id()) for object_id in object_ids
+            ]
+            timeout = timeout if timeout is not None else 2**30
```
duplicated

stephanie-wang commented on src/ray/object_manager/test/object_manager_test.cc:
```diff
@@ -236,6 +238,82 @@ class TestObjectManagerCommands : public TestObjectManager {
+    NextWaitTest();
+  }
+
+  void NextWaitTest() {
```
Can we separate each of the test cases into a separate unit test, and also use the gtest framework to be consistent with other raylet tests? Also, please document what case each of these is supposed to be testing.

elibol replied:
Yes, excellent point. I have a branch I'd like to merge before separating each test into a separate unit test as this is a somewhat substantial change.

Agree and will do the rest.

stephanie-wang commented on src/ray/object_manager/test/object_manager_test.cc:
```diff
@@ -236,6 +238,82 @@ class TestObjectManagerCommands : public TestObjectManager {
+                            const std::unordered_set<ray::ObjectID> &remaining) {
+          switch(current_wait_test){
+            case 0: {
+              ASSERT_TRUE(elapsed == 0);
```
Checking for `elapsed == 0` seems like it will be pretty flaky when run on a CI. Also, if we can, we should avoid exposing things that are only used for testing purposes as part of the API (e.g., exposing `elapsed` as an argument for the wait callback).

elibol replied:
The implementation explicitly sets elapsed to zero when `wait_ms` is 0, so this should never flake. Will look into whether it makes sense to remove elapsed.

stephanie-wang replied:
Okay, but I still think we should remove `elapsed`. We should write interfaces that we can unit test, but we should not write interfaces that are only for unit testing.

robertnishihara commented on src/local_scheduler/local_scheduler_client.cc:
```diff
@@ -186,3 +187,37 @@ void local_scheduler_set_actor_frontier(LocalSchedulerConnection *conn,
+    const std::vector<ObjectID> &object_ids,
+    int num_returns,
+    int64_t timeout,
+    bool wait_local) {
```
does `wait_local` work?

elibol replied:
No; I just added the piping for the argument. The back-end returns an unimplemented status if it's set to true. I have an idea of how to implement this whenever we'd like to add it.

robertnishihara commented on src/local_scheduler/local_scheduler_client.h:
```diff
@@ -169,4 +169,22 @@ const std::vector<uint8_t> local_scheduler_get_actor_frontier(
+/// \param wait_local Whether to wait for objects to appear on this node.
+/// \return A pair with the first element containing the object ids that were
+/// found,
+/// and the second element the objects that were not found.
```
looks like this could be moved to the previous line

robertnishihara commented on src/ray/object_manager/object_manager.h:
```diff
@@ -194,6 +195,30 @@ class ObjectManager : public ObjectManagerInterface {
   /// Cache of locally available objects.
   std::unordered_map<ObjectID, ObjectInfoT> local_objects_;
 
+  UniqueID object_directory_pull_callback_id_ = UniqueID::from_random();
```
`from_random` is expensive

robertnishihara replied:
oh I see this is only called once

robertnishihara replied:
why is this field needed?

robertnishihara commented on src/ray/raylet/format/node_manager.fbs:
```diff
@@ -53,7 +53,10 @@ enum MessageType:int {
+  ForwardTaskRequest,
+  // Wait for objects to be ready either from local or remote Plasma stores.
+  WaitRequest,
+  WaitReply
```
Document `WaitReply`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5732/
Test PASSed.

robertnishihara commented on src/ray/raylet/format/node_manager.fbs:
```diff
@@ -117,3 +120,21 @@ table ReconstructObject {
+  found: [string];
+  // List of object ids not found.
+  remaining: [string];
+}
```
should these actually go in `object_manager.fbs`?

robertnishihara commented on src/ray/raylet/node_manager.cc:
```diff
@@ -447,6 +447,36 @@ void NodeManager::ProcessClientMessage(
+          flatbuffers::Offset<protocol::WaitReply> wait_reply = protocol::CreateWaitReply(
+              fbb, to_flatbuf(fbb, found_vec), to_flatbuf(fbb, remaining_vec));
+          fbb.Finish(wait_reply);
+          client->WriteMessage(protocol::MessageType_WaitReply, fbb.GetSize(),
```
If the client dies in the middle of `client->WriteMessage`, are we handling that? Where will that error be handled?

elibol replied:
I added a check here; if the write fails an `IOError` status is returned.

robertnishihara commented on python/ray/worker.py:
```diff
@@ -2574,18 +2571,32 @@ def wait(object_ids, num_returns=1, timeout=None, worker=global_worker):
+        if worker.use_raylet:
+            if len(object_ids) != len(set(object_ids)):
+                raise Exception("Wait requires a list of unique object IDs.")
+            if len(object_ids) <= 0:
```
ah, it *should* be handled by the backend, but there's a bug and it was quicker to patch it here

https://github.com/ray-project/ray/issues/1969

robertnishihara replied:
on the topic of bugs, if you're wondering about the `2**30`, that's https://github.com/ray-project/ray/issues/58

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5740/
Test FAILed.

ericl commented:
@stephanie-wang @robertnishihara I want to add some color here. #1128 is an _extremely subtle and serious performance bug_ that is very hard for new (or even experienced) users to debug. It commonly manifests itself in a Ray cluster behaving as if it only has a single node.

It's quite easy to hit; in fact, RLlib has completely stopped using idiomatic `ray.wait` because of this problem.

That said, there a probably a few ways to solve it while preserving semantics. A couple acceptable ways of prioritizing returns would be:
1. random sampling
2. the order the user passed the ids in (if it's clear in the documentation)
3. the order the objects were sealed in

Unacceptable ways of prioritizing would be:
1. whatever order the object manager happened to return them in

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5742/
Test FAILed.

elibol commented:
@stephanie-wang the use of a UniqueID for ObjectDirectory Subscribe is IMHO a clean way to handle multiple subscriptions to the same object. There needs to be some way of identifying the listener callback, and this is one way to do it given the existing GCS API (which does not support subscribing to multiple times to the same object). I think we've used callbacks quite sparingly. Callbacks ought to be okay so long as we're not nesting them (i.e. callback hell).

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5744/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5746/
Test PASSed.

robertnishihara commented on python/ray/worker.py:
```diff
@@ -2574,18 +2571,31 @@ def wait(object_ids, num_returns=1, timeout=None, worker=global_worker):
+        if len(object_ids) != len(set(object_ids)):
+            raise Exception("Wait requires a list of unique object IDs.")
+        if len(object_ids) == 0:
+            raise Exception("Invalid number of objects %d." % len(object_ids))
```
Actually, this particular check should only be necessary for legacy Ray, and not for xray, right?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5788/
Test FAILed.

stephanie-wang commented:
@elibol, okay, since this is the way the rest of the object managt

@ericl, I see, I didn't realize this was a bug in `ray.wait`. Option 2 seems like it could be done without too much effort in this PR.

elibol commented:
@ericl thank you for elaborating. @stephanie-wang I agree option 2 is straight-forward in this PR.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5806/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5809/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5811/
Test FAILed.

elibol commented:
@ericl this now fixes #1128. It debiases local objects by doing a round of lookups for remote objects and allows the user to prioritize which objects are marked as ready by preserving the input order within both returned lists. For example, `ray.wait([x, y, z, w], num_returns=2) == ([y, z], [x, w])` if `x`, 'y', and `w` are ready. This is documented in `worker.py`.
@stephanie-wang I removed `elapsed` and refactored object directory to use a common helper function.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5810/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5813/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5814/
Test FAILed.

elibol commented:
jenkins, retest this please.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5816/
Test PASSed.

robertnishihara commented on src/ray/object_manager/object_manager.cc:
```diff
@@ -349,17 +350,132 @@ ray::Status ObjectManager::SendObjectData(const ObjectID &object_id,
+            if (!client_ids.empty()) {
+              wait_state.remaining.erase(object_id);
+              wait_state.found.insert(object_id);
+            }
```
What would it mean if `client_ids.empty() == true`? That would mean that the object has been evicted? How do we want to handle that case?

elibol replied:
It may mean that the object has been evicted. It may also mean that no entries exist for the object. In both cases, after the round of lookups, if we need to wait for more objects, `SubscribeObjectLocations` will be invoked on objects for which `client_ids.empty() == true` (in `AllWaitLookupsComplete`). The handler to `SubscribeObjectLocations` will be invoked when `!client_ids.empty()`.

robertnishihara commented on src/ray/object_manager/object_manager.cc:
```diff
@@ -349,17 +350,132 @@ ray::Status ObjectManager::SendObjectData(const ObjectID &object_id,
+              wait_state.remaining.erase(object_id);
+              wait_state.found.insert(object_id);
+            }
+            wait_state.requested_objects.erase(object_id);
```
It feels like removing the requested object even in the case where the object was not found could cause things to hang in some settings.

elibol replied:
In this pass, the `requested_objects` set is initialized to contain exactly the remaining objects. Each handler is invoked once. The last handler to be invoked is identified by `requested_objects.empty()`. A callback may not fire if GCS goes down, but then we have bigger problems. I don't think it's possible to hang here.

robertnishihara commented on src/ray/object_manager/object_manager.cc:
```diff
@@ -349,17 +350,132 @@ ray::Status ObjectManager::SendObjectData(const ObjectID &object_id,
+      RAY_CHECK_OK(object_directory_->LookupLocations(
+          oid, [this, wait_id](const std::vector<ClientID> &client_ids,
+                               const ObjectID &object_id) {
+            auto &wait_state = active_wait_requests_.find(wait_id)->second;
```
couldn't `find` not succeed at finding the `wait_id`?

elibol replied:
The only way to progress to a point where `active_wait_requests_.erase(wait_id)` is invoked is within `WaitLookupComplete`, which is only invoked when `requested_objects.empty()`, at which point no more handlers will be invoked with the same `wait_id`.

robertnishihara commented on src/ray/object_manager/object_manager.cc:
```diff
@@ -349,17 +350,132 @@ ray::Status ObjectManager::SendObjectData(const ObjectID &object_id,
+    }
+  }
+
+  if (wait_state.remaining.empty()) {
```
if we're going to include this special case here, then the condition should probably be something like `wait_state.found.size() >= num_required_objects || wait_ms == 0`.

That said, I think it makes sense to handle this case after firing off the `LookupLocations` calls.

elibol replied:
We invoke the lookup calls immediately after checking which objects are local to obtain current information about the location of remote objects. This allows us to collect information about all given objects, regardless of their location.

If we invoke when `wait_state.found.size() >= num_required_objects || wait_ms == 0` we may invoke `WaitComplete` before checking which remote objects are ready; we'll bias locally available objects.

If we fire `LookupLocations` and check `wait_state.found.size() >= num_required_objects || wait_ms == 0` independently, and the condition is `true`, the calls to `LookupLocations` will do nothing because the callback to `Wait` will already be invoked by the time any of the handlers for `LookupLocations` are invoked for the given `wait_id`.

robertnishihara commented on src/ray/object_manager/object_manager.cc:
```diff
@@ -349,17 +350,132 @@ ray::Status ObjectManager::SendObjectData(const ObjectID &object_id,
+    // TODO (hme): If we need to just wait for all objects independent of time
+    // (i.e. infinite wait time), determine what the value of wait_ms should be and
+    // skip this call. WaitComplete will be invoked when all objects have locations.
+    wait_state.timeout_timer->async_wait(
```
There may be some issues with the structure of the code.

For example, `Wait` will lead to one call to `WaitLookupComplete` for each requested object, and `WaitLookupComplete` will start the `timeout_timer`. But it only makes sense to start that timer once, (probably in `Wait`).

elibol replied:
`WaitLookupComplete` is invoked once for each invocation to `Wait`; I'll update the name of this method to `AllWaitLookupsComplete` make it easier to read.

robertnishihara commented on src/ray/object_manager/object_manager.cc:
```diff
@@ -349,17 +350,132 @@ ray::Status ObjectManager::SendObjectData(const ObjectID &object_id,
+    for (auto &oid : wait_state.remaining) {
+      // Subscribe to object notifications.
+      wait_state.requested_objects.insert(oid);
+      RAY_CHECK_OK(object_directory_->SubscribeObjectLocations(
```
Why do `SubscribeObjectLocations` here in addition to the `LookupLocations` in `Wait`? Wouldn't it make sense to just do one of those?

elibol replied:
The lookup calls return immediately, regardless of whether there's a known location of the object. After the lookup calls, if the conditions of the wait call are still not satisfied, we start the timer and subscribe to each remaining object. The subscription callback will be invoked only when an object's location is found. We wait until the minimum number of objects is found or the timeout is triggered.

elibol commented on src/ray/object_manager/object_manager.cc:
```diff
@@ -349,17 +350,132 @@ ray::Status ObjectManager::SendObjectData(const ObjectID &object_id,
+                                uint64_t num_required_objects, bool wait_local,
                                 const WaitCallback &callback) {
-  // TODO: Implement wait.
+  UniqueID wait_id = UniqueID::from_random();
```
Yes, I think it was a problem for `Pull`, when we were polling for locations with a low timeout. It's called once for each call to `Wait`, so I think it is okay. Is there an alternative to `from_random`? How are unique ObjectIDs created?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5820/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5818/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5821/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5826/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5827/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5828/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5835/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5836/
Test PASSed.

elibol commented on src/local_scheduler/local_scheduler_client.cc:
```diff
@@ -186,3 +187,37 @@ void local_scheduler_set_actor_frontier(LocalSchedulerConnection *conn,
+  // Read result.
+  int64_t type;
+  int64_t reply_size;
+  uint8_t *reply;
```
Added `free` at the end.

elibol commented on src/ray/object_manager/object_manager.cc:
```diff
@@ -349,17 +350,132 @@ ray::Status ObjectManager::SendObjectData(const ObjectID &object_id,
+    return ray::Status::NotImplemented("Wait for local objects is not yet implemented.");
+  }
+
+  RAY_CHECK(wait_ms >= 0);
```
I've allowed for `timeout_ms=-1` and added a test to ensure it's working properly (on the OM side).

stephanie-wang commented on src/ray/object_manager/object_manager.cc:
```diff
@@ -349,17 +350,132 @@ ray::Status ObjectManager::SendObjectData(const ObjectID &object_id,
+      RAY_CHECK_OK(object_directory_->SubscribeObjectLocations(
+          wait_id, oid, [this, wait_id](const std::vector<ClientID> &client_ids,
+                                        const ObjectID &object_id) {
+            auto &wait_state = active_wait_requests_.find(wait_id)->second;
```
There is an issue with the way this code is currently structured because of the fact that the callback registered in `SubscribeObjectLocations` may now get called directly. It's possible that right now we will not actually see a bug, but that is only because of the specific ordering of calls made on the object directory by the object manager, and that seems quite brittle. Here is a scenario where I think the code would break:

1. `SubscribeObjectLocations` gets called on objects A and B (e.g., for a `Pull`, or for a different `Wait`). Locations for both are cached in the object directory.
2. `Wait` is called on objects A and B, with 1 object required.
3. **In the same call stack**, `SubscribeObjectLocations` is called on object A. The cached locations are found, this callback fires, and the wait request completes and is erased from `active_wait_requests_`.
4. Again, **in the same call stack**, `SubscribeObjectLocations` is called on object B. `active_wait_requests_.find(wait_id)` will fail (silently).

stephanie-wang commented on test/runtest.py:
```diff
@@ -779,9 +779,6 @@ def testGetDict(self):
-    @unittest.skipIf(
-        os.environ.get("RAY_USE_XRAY") == "1",
-        "This test does not work with xray yet.")
     def testWait(self):
```
What happened to the other test cases that should now pass? I forget exactly why ones, but I think there was one like `testMultipleWaitsAndGets`?

elibol replied:
I've enabled all wait-related tests.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5871/
Test PASSed.

robertnishihara commented on src/local_scheduler/lib/python/local_scheduler_extension.cc:
```diff
@@ -179,6 +179,55 @@ static PyObject *PyLocalSchedulerClient_set_actor_frontier(PyObject *self,
+  int wait_local;
+
+  if (!PyArg_ParseTuple(args, "Oili", &py_object_ids, &num_returns, &timeout_ms,
+                        &wait_local)) {
```
I think the correct way to do this is to parse it as `O` and then use `PyObject_IsTrue`

elibol commented on src/ray/object_manager/object_manager.cc:
```diff
@@ -412,47 +417,58 @@ ray::Status ObjectManager::Wait(const std::vector<ObjectID> &object_ids, int64_t
-    for (auto &oid : wait_state.remaining) {
+    // Subscribe to objects in order to ensure Wait-related tests are deterministic.
+    for (auto &object_id : wait_state.object_id_order) {
+      if (wait_state.remaining.count(object_id) == 0) {
```
This is the relevant check that corrects the bug you described earlier: `active_wait_requests_.find(wait_id) == active_wait_requests_.end()`

stephanie-wang replied:
Yes, I don't think it will break in that particular way anymore, but undefined behavior is still possible because of the reference to `wait_state`. Same underlying issue, but it will break at a different line.

elibol replied:
Nice catch. I've made the fix along with a regression test.

elibol commented:
@stephanie-wang I can't think of a way to test the issue you're describing without adding test-specific code to Wait (e.g. skip lookup calls).

Before each `SubscribeObjectLocations` is issued, we check whether the executing wait request is still active. I believe this corrects the issue we discussed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5879/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5882/
Test PASSed.

stephanie-wang commented on src/ray/object_manager/object_manager.cc:
```diff
@@ -385,6 +397,12 @@ ray::Status ObjectManager::Wait(const std::vector<ObjectID> &object_ids,
     }
   }
 
+  return ray::Status::OK();
```
If this method always returns `OK`, I would make it `void`.

elibol replied:
It actually also returns `ray::Status::NotImplemented` currently if `wait_local=true`.

stephanie-wang commented on src/ray/object_manager/object_directory.h:
```diff
@@ -46,24 +46,39 @@ class ObjectDirectoryInterface {
   /// \param success_cb Invoked with non-empty list of client ids and object_id.
   /// \return Status of whether subscription succeeded.
-  virtual ray::Status SubscribeObjectLocations(const ObjectID &object_id,
+  virtual ray::Status SubscribeObjectLocations(const UniqueID &callback_id,
```
Can you add a `NOTE` to the documentation here that the callback may fire in the invocation of `SubscribeObjectLocations`? Until we can figure out a better way to do it, it'd be good to warn the user.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5894/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/5895/
Test PASSed.
